prow/cmd/branchprotector/protect.go

```diff
@@ -186,144 +164,96 @@ func (p *Protector) ConfigureBranches() {
-	for repo := range p.cfg.Presubmits {
-		if p.completedRepos[repo] == true {
-			continue
-		}
-		parts := strings.Split(repo, "/")
-		if len(parts) != 2 {
-			log.Fatalf("Bad repo: %s", repo)
-		}
-		orgName := parts[0]
-		repoName := parts[1]
-		repoReqs := repoRequirements(orgName, repoName, *p.cfg)
-		protect := len(repoReqs) > 0
-		if err := p.UpdateRepo(orgName, repoName, config.Repo{}, &protect, repoReqs, nil); err != nil {
-			p.errors.add(err)
-		}
-	}
-}
-
-// Update all repos in the org with the specified defaults
-func (p *Protector) UpdateOrg(orgName string, org config.Org, protect *bool, contexts, pushers []string) error {
-	if org.Protect != nil {
-		protect = org.Protect
-	}
-
-	oc := append([]string(nil), contexts...)
-	oc = append(oc, org.Contexts...)
-
-	op := append([]string(nil), pushers...)
-	op = append(op, org.Pushers...)
-
-	var repos []string
-	if protect != nil {
-		// Strongly opinionated org, configure every repo in the org.
-		rs, err := p.client.GetRepos(orgName, false)
-		if err != nil {
-			return fmt.Errorf("GetRepos(%s) failed: %v", orgName, err)
-		}
-		for _, r := range rs {
-			repos = append(repos, r.Name)
-		}
-	} else {
-		// Unopinionated org, just set explicitly defined repos
-		for r := range org.Repos {
-			repos = append(repos, r)
+	// Do not automatically protect tested repositories
+	if p.cfg.BranchProtection.ProtectTested {
+		for repo := range p.cfg.Presubmits {
+			parts := strings.Split(repo, "/")
+			if len(parts) != 2 {
+				log.Fatalf("Bad repo: %s", repo)
+			}
+			orgName := parts[0]
+			repoName := parts[1]
+			// Setting nil to fetch all the branches later.
+			tmpRefs[orgRepo{orgName, repoName}] = nil
 		}
 	}
 
-	for _, repoName := range repos {
-		err := p.UpdateRepo(orgName, repoName, org.Repos[repoName], protect, oc, op)
-		if err != nil {
-			return err
+	// Gathering org and repo that are configured in the Branch Protection settings
+	for org, orgP := range p.cfg.BranchProtection.Orgs {
+		if orgP.Protect != nil || p.cfg.BranchProtection.Protect != nil {
+			// Strongly opinionated org, configure every repo in the org.
+			gRepos, err := p.client.GetRepos(org, false)
+			if err != nil {
+				log.Printf("unable to get repos for org %s. %v", org, err)
+				p.errors.add(err)
+			}
+			for _, r := range gRepos {
+				// Setting nil to fetch all the branches later.
+				tmpRefs[orgRepo{org, r.Name}] = nil
+			}
+		} else {
+			for repo, repoP := range orgP.Repos {
+				if repoP.Protect != nil || orgP.Protect != nil || p.cfg.BranchProtection.Protect != nil {
+					// Setting to nil to fetch all the branches later.
+					tmpRefs[orgRepo{org, repo}] = nil
+				} else {
+					// If there is already an entry, then ProwJobs exist for that repo, and we need to fetch all branches
+					if _, exists := tmpRefs[orgRepo{org, repo}]; !exists {
+						var branches []string
+						for b := range repoP.Branches {
+							branches = append(branches, b)
```

Comment:
At the eighth level of indentation maybe we can break out some helper functions?